feat(mybatis): 优化 Mapper.xml 文件的热更新机制

loong95 opened the pull request:
- MyBatisSpringResourceManager保存Mapper文件的扫描路径（spring工程），用于管理 Mapper.xml 文件的扫描路径
- 在 MyBatisPlugin 中增加对 Mapper.xml 文件是否在配置路径中的检查
- 在 MyBatisSpringPatcher 中添加对 MyBatisProperties.resolveMapperLocations 方法的插桩，获取 mapperLocations
- 优化了 Mapper.xml 文件的加载逻辑，只加载在配置路径中的文件

loong95 committed:
feat(mybatis): 优化 Mapper.xml 文件的热更新机制

- MyBatisSpringResourceManager保存Mapper文件的扫描路径（spring工程），用于管理 Mapper.xml 文件的扫描路径
- 在 MyBatisPlugin 中增加对 Mapper.xml 文件是否在配置路径中的检查
- 在 MyBatisSpringPatcher 中添加对 MyBatisProperties.resolveMapperLocations 方法的插桩，获取 mapperLocations
- 优化了 Mapper.xml 文件的加载逻辑，只加载在配置路径中的文件

future0923 commented on debug-tools-hotswap/debug-tools-hotswap-plugin/debug-tools-hotswap-mybatis-plugin/src/main/java/io/github/future0923/debug/tools/hotswap/core/plugin/mybatis/patch/MyBatisSpringPatcher.java:
```diff
@@ -83,6 +83,24 @@ public static void patchClassPathMapperScanner(CtClass ctClass, ClassPool classP
+     * @param ctClass   ctClass
+     * @param classPool classPool
+     */
+    @OnClassLoadEvent(classNameRegexp = "org.mybatis.spring.boot.autoconfigure.MyBatisProperties|com.baomidou.mybatisplus.autoconfigure.MybatisProperties")
```
类名写错了，还有mybatis和mybatis plus的在两个patcher类中呢

<img width="1657" height="586" alt="image" src="https://github.com/user-attachments/assets/eefde5ef-0514-49e9-a889-7aaf7709e049" />

loong95 replied:
已修改

future0923 commented on debug-tools-hotswap/debug-tools-hotswap-plugin/debug-tools-hotswap-mybatis-plugin/src/main/java/io/github/future0923/debug/tools/hotswap/core/plugin/mybatis/patch/MyBatisSpringPatcher.java:
```diff
@@ -83,6 +83,24 @@ public static void patchClassPathMapperScanner(CtClass ctClass, ClassPool classP
+        try {
+            CtMethod resolveMapperLocations = ctClass.getDeclaredMethod("resolveMapperLocations");
+            resolveMapperLocations.insertAfter("{" +
+                    MyBatisSpringResourceManager.class.getName() + ".addMapperLocations(this.mapperLocations);" +
```
Javassist中`$_`可以拿到方法返回值，所以直接可以拿到方法的返回值  `Resource[] `，下面就不需要通过string[]在解析为Resource[]

loong95 replied:
这里如果拿这个返回值的话，就拿不到新增的Mapper.xml文件了，只能拿到启动时的文件

future0923 commented on debug-tools-hotswap/debug-tools-hotswap-plugin/debug-tools-hotswap-mybatis-plugin/src/main/java/io/github/future0923/debug/tools/hotswap/core/plugin/mybatis/reload/MyBatisSpringResourceManager.java:
```diff
@@ -107,4 +130,24 @@ public static ClassPathMapperScanner getMapperScanner() {
         return configurationList;
     }
+
+    public static boolean isInMapperLocations(String absolutePath) {
```
上面使用返回值拿到Resource[]，这就不需要每个xml都解析string[]为Resource[]在处理了，性能很低

future0923 commented on debug-tools-hotswap/debug-tools-hotswap-plugin/debug-tools-hotswap-mybatis-plugin/src/main/java/io/github/future0923/debug/tools/hotswap/core/plugin/mybatis/MyBatisPlugin.java:
```diff
@@ -97,9 +98,17 @@ public void watchResource(final URL url, final FileEvent fileEvent) throws Parse
         }
-        if ((FileEvent.CREATE.equals(fileEvent) && MyBatisUtils.isMapperXml(url.getPath()))
-                || ((FileEvent.MODIFY.equals(fileEvent) && configurationMap.containsKey(url.getPath())))) {
+        if ((FileEvent.CREATE.equals(fileEvent) && MyBatisUtils.isMapperXml(url.getPath()) && isInMapperLocations(url))
```
我感觉创建和修改类型都可以用这个判断一下生成不生成任务

loong95 replied:
> 我感觉创建和修改类型都可以用这个判断一下生成不生成任务

修改这里，我看原来已经有一个`configurationMap.containsKey(url.getPath())`的判断已经覆盖到了，所以是不用加了的

loong95 committed:
feat(mybatis): 支持 Spring Boot + MyBatis 和 MyBatis-Plus 工程的 Mapper.xml 热替换

- 在 MyBatisPlusPatcher 中添加对 MybatisPlusProperties.resolveMapperLocations 方法的插桩
- 在 MyBatisSpringPatcher 中添加对 MybatisProperties.resolveMapperLocations 方法的插桩
- 优化 MyBatisSpringResourceManager 中的资源比较逻辑，解决 Windows 上路径格式不一致的问题
